.gitlab-ci.yml

```diff
@@ -0,0 +1,27 @@
+stages:          # List of stages for jobs, and their order of execution
+  - build-docker
+  - test
+
+
+docker_build:
+  stage: build-docker
+  image: docker:20.10
+  script:
+    - DOCKER_HOST=tcp://localhost:2375 docker login -u $CI_REGISTRY_USER -p $CI_JOB_TOKEN $CI_REGISTRY
+    - docker build -t $CI_REGISTRY_IMAGE:$CI_BUILD_REF -f BaseDockerfile .
+    - docker push $CI_REGISTRY_IMAGE:$CI_BUILD_REF
+
+unit-tests:
+  stage: test
+  image: $CI_REGISTRY_IMAGE:$CI_BUILD_REF
+  variables:
+    CMAKE_ARGS: -DCUDA_ARCH=75
+  script:
+    - CMAKE_BUILD_PARALLEL_LEVEL=$(nproc) pip install -e .[dev,test]
```

Comment:
should we also `pip install -r requirements.txt` here?

Reply:
The dockerfile built in the build stage takes care of that

Reply:
Ah, I see. But I notice the dockerfile also contains `pip install -e .[dev,test]` -- by this logic, do we need to repeat this here? If we rebuild the image on every CI run, this seems redundant?

Reply:
We had discussed this offline, and for posterity: the docker build stage only targets the base environment. This way we have a fully fledged dockerfile in the repo, but only build and store what we need for CI.